pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/funnel/FunnelMaxStepAggregationFunction.java

```diff
@@ -0,0 +1,302 @@
+    long[] timestampBlock = blockValSetMap.get(_timestampExpression).getLongValuesSV();
+    List<int[]> stepBlocks = new ArrayList<>();
+    for (ExpressionContext stepExpression : _stepExpressions) {
+      stepBlocks.add(blockValSetMap.get(stepExpression).getIntValuesSV());
+    }
+    for (int i = 0; i < length; i++) {
+      for (int j = 0; j < _numSteps; j++) {
+        if (stepBlocks.get(j)[i] == 1) {
+          int[] groupKeys = groupKeysArray[i];
+          for (int groupKey : groupKeys) {
+            PriorityQueue<FunnelStepEvent> stepEvents = groupByResultHolder.getResult(groupKey);
+            if (stepEvents == null) {
+              stepEvents = new PriorityQueue<>(new FunnelStepEvent.FunnelStepEventComparator());
+            }
+            stepEvents.add(new FunnelStepEvent(timestampBlock[i], j));
+            groupByResultHolder.setValueForKey(groupKey, stepEvents);
+          }
+        }
+      }
+    }
+  }
+
+  @Override
+  public PriorityQueue<FunnelStepEvent> extractAggregationResult(AggregationResultHolder aggregationResultHolder) {
+    return aggregationResultHolder.getResult();
+  }
+
+  @Override
+  public PriorityQueue<FunnelStepEvent> extractGroupByResult(GroupByResultHolder groupByResultHolder, int groupKey) {
+    return groupByResultHolder.getResult(groupKey);
+  }
+
+  @Override
+  public PriorityQueue<FunnelStepEvent> merge(PriorityQueue<FunnelStepEvent> intermediateResult1,
+      PriorityQueue<FunnelStepEvent> intermediateResult2) {
+    if (intermediateResult1 == null) {
+      return intermediateResult2;
+    }
+    if (intermediateResult2 == null) {
+      return intermediateResult1;
+    }
+    intermediateResult1.addAll(intermediateResult2);
+    return intermediateResult1;
+  }
+
+  @Override
+  public DataSchema.ColumnDataType getIntermediateResultColumnType() {
+    return DataSchema.ColumnDataType.OBJECT;
+  }
+
+  @Override
+  public DataSchema.ColumnDataType getFinalResultColumnType() {
+    return DataSchema.ColumnDataType.LONG;
+  }
+
+  @Override
+  public Long extractFinalResult(PriorityQueue<FunnelStepEvent> stepEvents) {
+    long finalMaxStep = 0;
+    if (stepEvents == null || stepEvents.isEmpty()) {
+      return finalMaxStep;
+    }
+    ArrayDeque<FunnelStepEvent> slidingWindow = new ArrayDeque<>();
+    while (!stepEvents.isEmpty()) {
+      fillWindow(stepEvents, slidingWindow);
+      if (slidingWindow.isEmpty()) {
+        break;
+      }
+      int maxSteps = processWindow(slidingWindow);
+      finalMaxStep = Math.max(finalMaxStep, maxSteps);
+      if (finalMaxStep == _numSteps) {
+        break;
+      }
+      long windowStart = slidingWindow.peek().getTimestamp();
+      while (!slidingWindow.isEmpty() && slidingWindow.peek().getTimestamp() == windowStart) {
+        slidingWindow.pollFirst();
+      }
+    }
+    return finalMaxStep;
+  }
+
+  private void fillWindow(PriorityQueue<FunnelStepEvent> stepEvents, ArrayDeque<FunnelStepEvent> slidingWindow) {
```

Comment:
Is this algorithm correct? Do we need to calculate all windows starts with step 0?
Currently we will miss all step 0 falling in the window with the previous step 0

Reply:
No, the sliding window will process to the next step0 if it's inside the window. Or it will pop all the events and insert from stepEvents.